integration/tests/access/unstaked_node_test.go

```diff
@@ -41,108 +41,76 @@ func TestUnstakedAccessSuite(t *testing.T) {
-	storage := node1.Storage
-	require.NotNil(suite.T(), storage)
-	blocks := storage.Blocks
-	require.NotNil(suite.T(), blocks)
-
-	rcvdBlockCnt := 0
-	var err error
-	receivedBlocks := make([]flow.Identifier, blockCount)
-	receiveBlocks := func() {
-		for ; rcvdBlockCnt < blockCount; rcvdBlockCnt++ {
-			select {
-			case blockID := <-suite.finalizedBlockIDsChans[0]:
-				receivedBlocks[rcvdBlockCnt] = blockID
-				_, err = blocks.ByID(blockID)
-				if err != nil {
-					return
+	receivedBlocks := make(map[flow.Identifier]struct{}, blockCount)
+
+	suite.Run("consensus follower follows the chain", func() {
+
+		// kick off the first follower
+		suite.followerMgr1.startFollower(ctx)
+		var err error
+		receiveBlocks := func() {
+			for i := 0; i < blockCount; i++ {
+				select {
+				case blockID := <-suite.followerMgr1.blockIDChan:
+					receivedBlocks[blockID] = struct{}{}
+					_, err = suite.followerMgr1.hasBlock(blockID)
+					if err != nil {
+						return
+					}
 				}
 			}
 		}
-	}
-
-	// wait for finalized blocks
-	unittest.AssertReturnsBefore(suite.T(), receiveBlocks, 2*time.Minute) // waiting 1 minute for 5 blocks
 
-	// all blocks were found in the storage
-	require.NoError(suite.T(), err, "finalized block not found in storage")
+		// wait for finalized blocks
+		unittest.AssertReturnsBefore(suite.T(), receiveBlocks, 2*time.Minute) // waiting 2 minute for 5 blocks
 
-	// assert that blockCount number of blocks were received
-	require.Equal(suite.T(), blockCount, rcvdBlockCnt)
+		// all blocks were found in the storage
+		require.NoError(suite.T(), err, "finalized block not found in storage")
 
-	// kick off the second follower
-	go suite.followers[1].Run(ctx)
+		// assert that blockCount number of blocks were received
+		require.Len(suite.T(), receivedBlocks, blockCount)
+	})
 
-	followerImpl2, ok := suite.followers[1].(*consensus_follower.ConsensusFollowerImpl)
-	if !ok {
-		suite.Fail("unexpected consensus follower implementation")
-		return
-	}
+	suite.Run("consensus follower sync up with the chain", func() {
 
-	// get the underlying node builder
-	node2 := followerImpl2.NodeBuilder
-	// wait for the follower to have completely started
-	unittest.RequireCloseBefore(suite.T(), node2.Ready(), 10*time.Second,
-		"timed out while waiting for consensus follower to start")
+		// kick off the second follower
+		suite.followerMgr2.startFollower(ctx)
 
-	receiveBlocks2 := func() {
-		for _, blockID := range receivedBlocks {
-			select {
-			case receivedBlockID := <-suite.finalizedBlockIDsChans[1]:
-				suite.Assert().Equal(blockID, receivedBlockID)
+		// the second follower is now atleast blockCount blocks behind and should sync up and get all the missed blocks
+		receiveBlocks := func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case blockID := <-suite.followerMgr2.blockIDChan:
+					delete(receivedBlocks, blockID)
+					if len(receivedBlocks) == 0 {
+						return
+					}
+				}
```

Comment:
Hmm, what is the benefit of doing it this way compared to using an array?

Is it because of possible race condition?

Reply:
benefit was that is makes it easy to just mark off the block IDs that have been received. Array would make require a `find-then-delete` pattern.
The next block ID will be pulled from the channel only after the current one has been deleted or ignored since this loop is run by single thread.